sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala

```diff
@@ -137,3 +149,53 @@ object PropagateEmptyRelation extends Rule[LogicalPlan] with PredicateHelper wit
     }
   }
 }
+
+/**
+ * Rule [[PropagateEmptyRelationBase]] at normal optimizer side.
```

Comment:
let's also mention why we don't do it in the AQE side: these optimizations are not signigicant and may introduce extra exchanges.

Reply:
Updated.